pypesto/ensemble/covariance_analysis.py

```diff
@@ -0,0 +1,318 @@
+        above of this cutoff
+        (only applied when only_identifiable_directions is True)
+
+    inv_upper_relative_cutoff:
+        Consider only low eigenvalues of the covariance matrix when rescaled
+        with the largest eigenvalue with inverses above of this cutoff
+        (only applied when only_identifiable_directions is True)
+
+    only_identifiable_directions:
+        return only identifiable directions according to inv_upper_cutoff
+
+    only_separable_directions:
+        return only separable directions according to lower_cutoff
+
+    Returns
+    =======
+    eigen_values:
+        Eigenvalues of the covariance matrix
+
+    eigen_vectors:
+        eigenvectors of the covariance matrix
+    """
+
+    # get the eigenvalue decomposition
+    eigen_vals, eigen_vectors = np.linalg.eigh(matrix_to_analyze)
+
+    # get a normalized version
+    rel_eigen_vals = eigen_vals / np.max(eigen_vals)
+
+    # If no filtering is wanted, we can return
+    if not only_identifiable_directions and not only_separable_directions:
+        # apply normlization
+        if normalize:
+            eigen_vals = rel_eigen_vals
+        return eigen_vals, eigen_vectors
+
+    if only_identifiable_directions and only_identifiable_directions:
+        raise Exception('Asking for only identiafiable and only separable '
+                        'directions at the same time makes no sense. The '
+                        'applied filters are mutually exclusive.')
+
+    # Separable directions are wanted: an upper pass filtering is needed
+    if only_separable_directions:
+        if lower_absolute_cutoff is not None and \
+                lower_relative_cutoff is not None:
+            above_cutoff = np.array([
+                eigen_vals[i_eig] > lower_absolute_cutoff and
+                rel_eigen_vals[i_eig] > lower_relative_cutoff
+                for i_eig in range(len(eigen_vals))
+            ])
+        elif lower_absolute_cutoff is not None:
+            above_cutoff = eigen_vals > lower_absolute_cutoff
+        elif lower_relative_cutoff is not None:
+            above_cutoff = rel_eigen_vals > lower_relative_cutoff
+        else:
+            raise Exception('Need a lower cutoff (aboslute or relative, '
+                            'e.g., 1e-15, to compute separable directions.')
+
+        # restrict to those above cutoff
+        eigen_vals = eigen_vals[above_cutoff]
+        eigen_vectors = eigen_vectors[:, above_cutoff]
+        # apply normlization
+        if normalize:
+            eigen_vals = rel_eigen_vals
+        return eigen_vals, eigen_vectors
+
+    # Identifiable directions are wanted: an filtering of the inverse
+    # eigenvalues is needed (upper pass of inverse = lower pass of original)
+    if inv_upper_absolute_cutoff is not None and \
+            inv_upper_relative_cutoff is not None:
+        below_cutoff = np.array([
+            eigen_vals[i_eig] < 1 / eigen_vals > inv_upper_absolute_cutoff and
```

Comment:
Should `1 / eigen_vals > inv_upper_absolute_cutoff` be checked each time?

Reply:
thanks, was an error!
Need to add a test...